core/src/main/java/org/hisp/dhis/android/core/domain/metadata/MetadataCall.kt

```diff
@@ -96,80 +99,71 @@ internal class MetadataCall @Inject constructor(
 ) {
 
     companion object {
-        const val CALLS_COUNT = 12
+        const val CALLS_COUNT = 15
     }
 
-    fun download(): Observable<D2Progress> {
+    @Suppress("TooGenericExceptionCaught")
+    fun download(): Flow<D2Progress> = channelFlow {
         val progressManager = D2ProgressManager(CALLS_COUNT)
-        return changeEncryptionIfRequired().andThen(
-            rxCallExecutor.wrapObservableTransactionally(
-                Observable.merge(
-                    systemInfoDownloader.downloadWithProgressManager(progressManager),
-                    executeIndependentCalls(progressManager),
-                    executeUserCallAndChildren(progressManager)
-                ),
-                true
-            )
-        )
+        changeEncryptionIfRequired().blockingAwait()
+        coroutineAPICallExecutor.wrapTransactionally(cleanForeignKeyErrors = true) {
+            try {
+                systemInfoDownloader.downloadWithProgressManager(progressManager).also { send(it) }
+                executeIndependentCalls(progressManager).collect { send(it) }
+                executeUserCallAndChildren(progressManager).collect { send(it) }
+            } catch (e: Exception) {
```

Comment:
This catch block is required to unwrap the exception thrown by the RxJava methods (see below). The methods contains RxJava statements such as `.blockingAwait()`, which wraps the D2Error exception in a RuntimeException.
Once all the queries are refactored to coroutines, this catch block could be removed.